### PR TITLE
Create Table and use flyway Migration

### PR DESCRIPTION
**Pull request by Usha3012**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,14 @@
...
     ToDo's :
-    Create new branch named as : Issue-1-spring-init
+    
+    Add Spring JPA Hibernate dependencies to pom.xml
+    Use flyway to generate database and schema.
+    Create schema goal .
+    Create table Category
+    Create package entity in src/main/java
+    Create Entity class Category
+
```
Comment (njoysubho) [inline README.md]: This looks very bad .. Remove this from Readme maintain it somewhere else

File: src/main/java/org/api/entity/Category.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.sql.Timestamp;
+import java.util.UUID;
+
+@Entity
+@Data
+@AllArgsConstructor
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java:17]: Hibernate also need a NoArgs constructor

File: src/main/java/org/api/entity/Category.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+@Table(name = "Category" ,schema = "goal")
+public class Category {
+
+    @Id
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java:22]: This should be UUID with autogeneration

File: src/main/java/org/api/entity/Category.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    private  Integer id;
+
+    @Column(name = "category_name")
+    private String category_name;
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java]: Please follow camelCase naming convention for all properties

File: src/main/java/org/api/entity/Category.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Column(name = "created_by")
+    private UUID created_by;
+
+    @Column(name = "created_on")
+    private Timestamp created_on;
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java]: created_on and modified_on should be auto populated by hibernate. Use @CreatedDate and @LastModifedDate.
Bonus point - Try to understand how hibernate actually process those annotation.

File: src/main/java/org/api/entity/Category.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Column(name = "system_defined")
+    private Boolean system_defined;
+
+    @Column(name = "modified_on")
+    private  Timestamp modified_on;
+
+    @Column(name = "tags")
+    private String tags;
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java:45]: We may need to revisit this what if we have more than one tags. May be we can keep comma separated string for now

File: src/main/resources/db/migration/V1__base.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
+SET search_path = goal, pg_catalog;
```
Comment (njoysubho) [inline src/main/resources/db/migration/V1__base.sql:2]: where is the schema ??

Always write migration code as rerunnable check how to use IF NOT EXISTS with create table and schema

File: src/main/resources/db/migration/V1__base.sql [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+CREATE TABLE Category (
+    id SERIAL PRIMARY KEY,
+    category_name VARCHAR,
```
Comment (njoysubho) [inline src/main/resources/db/migration/V1__base.sql]: what is the length of varchar

File: src/main/resources/db/migration/V1__base.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
+SET search_path = goal, pg_catalog;
+
+CREATE TABLE Category (
+    id SERIAL PRIMARY KEY,
```
Comment (njoysubho) [inline src/main/resources/db/migration/V1__base.sql]: No we do not need SERIAL make it UUID

File: src/main/resources/db/migration/V1__base.sql [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+    category_name VARCHAR,
+    created_by UUID ,
+    created_on TIMESTAMP,
+    system_defined BOOLEAN,
```
Comment (njoysubho) [inline src/main/resources/db/migration/V1__base.sql]: use default value false

**Review by njoysubho — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Usha3012) [reply README.md]: Removed

File: src/main/java/org/api/entity/Category.java (hunk 3 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java:22]: Changed from Integer to UUID with autogeneration

File: src/main/java/org/api/entity/Category.java (hunk 4 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java]: Done

File: src/main/java/org/api/entity/Category.java (hunk 6 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java:45]: Not clear to me , what you want to convey ?

File: src/main/java/org/api/entity/Category.java (hunk 5 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java]: Done

File: src/main/resources/db/migration/V1__base.sql (hunk 7 above)
Comment (Usha3012) [reply src/main/resources/db/migration/V1__base.sql:2]: Done

File: src/main/resources/db/migration/V1__base.sql (hunk 8 above)
Comment (Usha3012) [reply src/main/resources/db/migration/V1__base.sql]: Done

File: src/main/resources/db/migration/V1__base.sql (hunk 9 above)
Comment (Usha3012) [reply src/main/resources/db/migration/V1__base.sql]: Done

File: src/main/resources/db/migration/V1__base.sql (hunk 10 above)
Comment (Usha3012) [reply src/main/resources/db/migration/V1__base.sql]: Done

File: src/main/java/org/api/entity/Category.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+@Entity
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Table(name = "Category" ,schema = "goal")
+public class Category {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java]: uuid needs a different generator

File: src/main/java/org/api/entity/Category.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    private  UUID id;
+
+    @Column(name = "category_name")
+    private String category_Name;
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java]: where is cameCase ?? 🤒

File: src/main/java/org/api/entity/Category.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    private Timestamp created_On;
+
+    @Column(name = "system_defined")
+    private Boolean system_Defined;
+
+    @LastModifiedDate
+    @Column(name = "modified_on")
+    private  Timestamp modified_On;
+
+    @Column(name = "tags")
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java]: CAMELCASE

File: src/main/resources/db/migration/V1__base.sql [hunk 14]
```diff
@@ -0,0 +1,12 @@
+create schema if not exists goal
+SET search_path = goal, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS goal.Category(
+    id UUID PRIMARY KEY,
+    category_name VARCHAR(255),
```
Comment (njoysubho) [inline src/main/resources/db/migration/V1__base.sql:6]: make it not nullable

**Review by njoysubho — CHANGES_REQUESTED** [review]
Not all comments addresses

File: src/main/java/org/api/entity/Category.java (hunk 13 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java]: Got it :(

File: src/main/java/org/api/entity/Category.java (hunk 11 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java]: GenerationType.AUTO 
Done

File: src/main/java/org/api/entity/Category.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+@Entity
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Table(name = "Category" ,schema = "goal")
+public class Category {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private  UUID id;
```
Comment (njoysubho) [inline src/main/java/org/api/entity/Category.java:25]: this will not work check how to use uuid generator

**Review by njoysubho — CHANGES_REQUESTED** [review]
one more issue to go

File: src/main/java/org/api/entity/Category.java (hunk 15 above)
Comment (Usha3012) [reply src/main/java/org/api/entity/Category.java:25]: Done
@GeneratedValue(generator = "uuid2")
@GenericGenerator(name = "uuid2", strategy = "uuid2")

**Review by njoysubho — APPROVED** [review]

**Pull request merged**